### PR TITLE
Random

### LINKED ISSUE #488: kiwix-serve random is not seeded properly on Windows
*@OscarL commented on Jan 31, 2021, 5:17 PM UTC:*

I keep getting the same articles so frequently, and in the same order, that I made it to a memory game... (its "0..999" or "Macao Special Administrative Region passport" that comes next?) :-)

Happens with an old version of kiwix-serve.exe from 2019, and with 3.1.2 downloaded today. ~~In fact, the newer version is even worse, as it tends to repeat the same articles much more frequently, and if you hit the dice "too fast", it tends to hang the server.~~ [#436](https://github.com/kiwix/kiwix-tools/issues/436)

Happens with Edge and with Firefox (slightly less bad on Firefox it seems, but maybe its because I tried it on Edge first).

Does not happens when running kiwix-serve (2019 version) from a Linux install.

So I guess the Windows version of the random function is not up to par, but at least should be seeded properly :-)

*This issue was moved by [kelson42](https://github.com/kelson42) from [kiwix/kiwix-tools#435](https://github.com/kiwix/kiwix-tools/issues/435).*

### PR DESCRIPTION
**Pull request by mgautierfr**
Need merge of ~~#466~~ #487 first.

Fixes #488

### CODE REVIEW

**Comment by kelson42** [comment]
@mgautierfr will that fix https://github.com/openzim/warc2zim/issues/37 ?

File: src/archive.cpp [hunk 1]
```diff
@@ -178,6 +178,23 @@ namespace zim
...
   }
 
+  Entry Archive::getRandomEntry() const {
+    auto frontEntryCount = m_impl->getFrontEntryCount().v;
+    int watchdog = 42;
+    while(--watchdog) {
+      auto idx = ((double)rand() / ((double)RAND_MAX+1)) * frontEntryCount;
+      auto entry = getEntryByTitle(idx);
+      auto item = entry.getItem(true);
+      std::cerr << item.getPath() << "=>" << item.getMimetype() << std::endl;
```
Comment (veloman-yunkan) [inline src/archive.cpp:188]: Do we need this bare `std::cerr` here? Maybe wrap it in a logging statement?

File: test/archive.cpp [hunk 2]
```diff
@@ -162,6 +162,25 @@ TEST(ZimArchive, openRealZimArchive)
...
+  };
+
+  for ( const std::string fname : zimfiles ) {
+    const auto path = zim::DEFAULTFS::join("data", fname);
+    const TestContext ctx{ {"path", path } };
+    const zim::Archive archive(path);
+    const auto randomEntry = archive.getRandomEntry();
+    const auto item = randomEntry.getItem(true);
+    ASSERT_TRUE(item.getMimetype().find("text/html") != std::string::npos) << ctx;
+  }
```
Comment (veloman-yunkan) [inline test/archive.cpp:181]: It also makes sense to call `getRandomEntry()` several times and check that its result is indeed random. To this end `small.zim` will need to be either excluded from this test, or enhanced with more pages (currently it has only one front page).

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]
I reviewed only the last commit. In fact I wonder, why this PR doesn't have base branch set to `indexes`. Then only the last commit will show up in this PR and the base branch will be automatically reset to `master` when `indexes` is merged.

**Comment by kelson42** [comment]
@mgautierfr Would be an good opportunity to fix #488 here IMO.

**Comment by mgautierfr** [comment]
The last push is a better implementation of random :
- Use cpp random version
- Make the random thread safe (rand is not)
- Properly initialize random with time as seed (should fix #488)

File: src/archive.cpp (hunk 1 above)
Comment (mgautierfr) [reply src/archive.cpp]: I've removed the line.

File: test/archive.cpp (hunk 2 above)
Comment (mgautierfr) [reply test/archive.cpp]: Done

File: src/archive.cpp [hunk 3]
```diff
@@ -184,6 +184,22 @@ namespace zim
...
+      auto idx = randomNumber(frontEntryCount);
+      auto entry =  getEntryByTitle(idx);
+      auto item = entry.getItem(true);
+
+      if (item.getMimetype().find("text/html") == std::string::npos) {
+        continue;
+      }
+      return entry;
+    }
+    throw EntryNotFound("Cannot found valid random entry");
```
Comment (veloman-yunkan) [inline src/archive.cpp:200]: found → find

File: include/zim/archive.h [hunk 4]
```diff
@@ -212,6 +212,16 @@ namespace zim
        */
       Entry getFaviconEntry() const;
 
+
+      /** Get a random entry.
+       *
+       * The entry is pick randomly from the front artice list.
+       *
+       * @return A random entry.
+       * @exception EntryNoFound If no valid random entry can be found.
```
Comment (veloman-yunkan) [inline include/zim/archive.h:221]: EntryNoFound → EntryNotFound

File: src/fileimpl.h [hunk 5]
```diff
@@ -111,7 +111,11 @@ namespace zim
 
       entry_index_t getStartUserEntry() const { return m_startUserEntry; }
       entry_index_t getEndUserEntry() const { return m_endUserEntry; }
+      // The number of entry added by the creator. (So excluding index, ...).
+      // On new namespace scheme, number of entry in C namespace
```
Comment (veloman-yunkan) [inline src/fileimpl.h:115]: entry → entries

File: src/fileimpl.h [hunk 6]
```diff
@@ -111,7 +111,11 @@ namespace zim
...
       entry_index_t getUserEntryCount() const { return m_endUserEntry - m_startUserEntry; }
+      // The number of entry that can be considered as front article (no resource)
```
Comment (veloman-yunkan) [inline src/fileimpl.h:117]: entry → entries

File: src/tools.cpp [hunk 7]
```diff
@@ -114,3 +116,13 @@ std::tuple<char, std::string> zim::parseLongPath(const std::string& longPath)
...
 }
+
+uint32_t zim::randomNumber(uint32_t max)
+{
+  static std::default_random_engine random(
+    std::chrono::system_clock::now().time_since_epoch().count());
+  static std::mutex mutex;
+
+  std::lock_guard<std::mutex> l(mutex);
+  return ((double)random() / RAND_MAX) * max;
```
Comment (veloman-yunkan) [inline src/tools.cpp:127]: use `random.max()` instead of `RAND_MAX`

File: src/tools.h [hunk 8]
```diff
@@ -34,6 +34,12 @@ namespace zim {
   void microsleep(int microseconds);
 
   std::tuple<char, std::string> parseLongPath(const std::string& longPath);
+
+  /** Return a random number between [0, max]
```
Comment (veloman-yunkan) [inline src/tools.h:38]: between → from the range

File: src/archive.cpp [hunk 9]
```diff
@@ -184,6 +184,22 @@ namespace zim
     }
   }
 
+  Entry Archive::getRandomEntry() const {
+    auto frontEntryCount = m_impl->getFrontEntryCount().v;
+    int watchdog = 42;
+    while(--watchdog) {
+      auto idx = randomNumber(frontEntryCount);
```
Comment (veloman-yunkan) [inline src/archive.cpp:191]: `frontEntryCount` → `frontEntryCount-1` (and handle the case of an empty ZIM archive)

File: test/archive.cpp [hunk 10]
```diff
@@ -153,6 +153,33 @@ TEST(ZimArchive, openRealZimArchive)
...
+    const zim::Archive archive(path);
+    auto randomEntry = archive.getRandomEntry();
+    const auto item = randomEntry.getItem(true);
+    ASSERT_TRUE(item.getMimetype().find("text/html") != std::string::npos) << ctx;
+    if (fname == "small.zim" ) {
+      continue;
+    }
+    for (auto i = 0; i<10; i++) {
+      const auto newRandomEntry = archive.getRandomEntry();
+      ASSERT_NE(randomEntry.getIndex(), newRandomEntry.getIndex());
```
Comment (veloman-yunkan) [inline test/archive.cpp:177]: This can occasionally fail

File: test/random.cpp [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+TEST(Random, pseudoUnique)
+{
+  auto prev = randomNumber(1000000);
+  for (auto i=0; i<1000; i++) {
+    auto r = randomNumber(1000000);
+    ASSERT_NE(prev, r);
+    prev = r;
+  }
```
Comment (veloman-yunkan) [inline test/random.cpp:50]: An alternating sequence 0, 1, 0, 1, ... returned by `randomNumber` will pass this test. On the other hand, in theory this unit-test may occasionally fail (for a good enough random number generator). But that is a universal challenge for creating a good unit test for random number generators. Why don't you look-up for an existing unit test on the internet and adapt it to your function?

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]

File: src/archive.cpp (hunk 9 above)
Comment (mgautierfr) [reply src/archive.cpp]: Good catch

**Comment by mgautierfr** [comment]
I've fixed your comments and add a better test on the `randomNumber` function.

But, honestly I think it is pretty hard to test randomness. I've added the tests because you ask for them.
I haven't found a existing unit-test.
Best comments on internet said that unit test is not a good test for random function.
Some others say we should test the distribution of the results (but it is not enough).

Here I'm felling as testing the std random method. It breaks the black-box principle of writing test but the only reason the output of `randomNumber` is not properly distribute is that std random is broken.
The random number is just use to get a random page. There is nothing critical in the use (especially security).

A careful review would be better than trying to do a complete unit test on a random method.

File: test/random.cpp [hunk 12]
```diff
@@ -36,17 +36,30 @@ TEST(Random, smallMax)
...
+  const uint32_t NB_NUMBERS = 1000000;
+  const uint32_t NB_BUCKETS = 100;
+  const uint32_t BUCKET_SIZE = NB_NUMBERS/NB_BUCKETS;
+  const uint32_t MAX_RANDOM = 1000000;
+  std::vector<uint32_t> distribution(NB_BUCKETS);
+
+  for (auto i=0U; i<NB_NUMBERS; i++) {
+    auto r = randomNumber(MAX_RANDOM);
+    auto bucket_index = (float)r / MAX_RANDOM * NB_BUCKETS;
+    distribution[bucket_index]++;
```
Comment (veloman-yunkan) [inline test/random.cpp:54]: `bucket_index` can be equal to `NB_BUCKETS`

File: src/fileimpl.h [hunk 13]
```diff
@@ -112,9 +112,9 @@ namespace zim
       entry_index_t getStartUserEntry() const { return m_startUserEntry; }
       entry_index_t getEndUserEntry() const { return m_endUserEntry; }
       // The number of entry added by the creator. (So excluding index, ...).
```
Comment (veloman-yunkan) [inline src/fileimpl.h:114]: On this line entry was not converted to plural

File: include/zim/archive.h [hunk 14]
```diff
@@ -218,7 +218,7 @@ namespace zim
        * The entry is pick randomly from the front artice list.
```
Comment (veloman-yunkan) [inline include/zim/archive.h:218]: pick → picked

File: src/archive.cpp [hunk 15]
```diff
@@ -188,7 +188,7 @@ namespace zim
     auto frontEntryCount = m_impl->getFrontEntryCount().v;
     int watchdog = 42;
     while(--watchdog) {
-      auto idx = randomNumber(frontEntryCount);
+      auto idx = randomNumber(frontEntryCount-1);
       auto entry =  getEntryByTitle(idx);
       auto item = entry.getItem(true);
```
Comment (veloman-yunkan) [inline src/archive.cpp:193]: The (not so important) corner case of an empty ZIM archive (or, rather, a ZIM archive without any front entries) remained unhandled.

**Review by veloman-yunkan — CHANGES_REQUESTED** [review]
I was about to approve this PR, and only listed some more minor issues because of a more or less serious issue with `NB_BUCKETS` in the `Random.distribution` unit test.

**Comment by kelson42** [comment]
@mgautierfr This PR is clearly the next one in the pipe, it seems there a few things which still need to be fixed.

**Comment by mgautierfr** [comment]
Fix issues found by @veloman-yunkan.
Add better message if random test is failing (which may not be a bug).

**Review by veloman-yunkan — APPROVED** [review]

**Comment by mgautierfr** [comment]
I've removed small.zim from the random unit test as it tends to fail a bit too much.

**Pull request merged**